<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="UTF-8">

<!-- Begin Jekyll SEO tag v2.8.0 -->
<title>Introduction to RNA-seq using high-performance computing (January 2025) | Branched Docs Demo</title>
<meta name="generator" content="Jekyll v3.10.0" />
<meta property="og:title" content="Introduction to RNA-seq using high-performance computing (January 2025)" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Example of building HTML docs hosted on github.io with subdirectories for selected branches" />
<meta property="og:description" content="Example of building HTML docs hosted on github.io with subdirectories for selected branches" />
<link rel="canonical" href="https://nichd-bspc.github.io/branched-docs-demo/site-v2/" />
<meta property="og:url" content="https://nichd-bspc.github.io/branched-docs-demo/site-v2/" />
<meta property="og:site_name" content="Branched Docs Demo" />
<meta property="og:type" content="website" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="Introduction to RNA-seq using high-performance computing (January 2025)" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"WebSite","description":"Example of building HTML docs hosted on github.io with subdirectories for selected branches","headline":"Introduction to RNA-seq using high-performance computing (January 2025)","name":"Branched Docs Demo","url":"https://nichd-bspc.github.io/branched-docs-demo/site-v2/"}</script>
<!-- End Jekyll SEO tag -->

    <link rel="preconnect" href="https://fonts.gstatic.com">
    <link rel="preload" href="https://fonts.googleapis.com/css?family=Open+Sans:400,700&display=swap" as="style" type="text/css" crossorigin>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="theme-color" content="#157878">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    <link rel="stylesheet" href="/branched-docs-demo/site-v2/assets/css/style.css?v=d9537db2916b06ce862a1b3ed98634ec415b0bc4">
    <!-- start custom head snippets, customize with your own _includes/head-custom.html file -->

<!-- Setup Google Analytics -->



<!-- You can set your favicon here -->
<!-- link rel="shortcut icon" type="image/x-icon" href="/branched-docs-demo/site-v2/favicon.ico" -->

<!-- end custom head snippets -->

  </head>
  <body>
    <a id="skip-to-content" href="#content">Skip to the content.</a>

    <header class="page-header" role="banner">
      <h1 class="project-name">Introduction to RNA-seq using high-performance computing (January 2025)</h1>
      <h2 class="project-tagline">Example of building HTML docs hosted on github.io with subdirectories for selected branches</h2>
      
        <a href="https://github.com/NICHD-BSPC/branched-docs-demo" class="btn">View on GitHub</a>
      
      
    </header>

    <main id="content" class="main-content" role="main">
      <h1 id="introduction-to-rna-seq-using-high-performance-computing-january-2025">Introduction to RNA-seq using high-performance computing (January 2025)</h1>

<h3 id="description">Description</h3>

<p>This repository has teaching materials for a 9-week Introduction to RNA-sequencing data analysis workshop held at NICHD from Jan-March 2025. This workshop focuses on teaching basic computational skills to enable the effective use of an high-performance computing environment to implement an RNA-seq data analysis workflow. The first four and a half weeks includes an introduction to shell (bash) and shell scripting, running the RNA-seq workflow from FASTQ files to count data and covers best practice guidelines for RNA-seq experimental design and data organization/management. The latter half covers differential expression analyses using DEseq2 and downstream analyses of those results.</p>

<h3 id="learning-objectives">Learning Objectives</h3>

<ul>
  <li>Gain practical knowledge about analyzing RNAseq from experimental design through functional enrichment analysis</li>
  <li>Learn broadly applicable bioinformatics skills such as command line and R programming</li>
  <li>Work with real data sets and real bioinformatics environments on NIH’s high-performance compute cluster (Biowulf)</li>
  <li>Apply what you learn to your own bulk RNAseq data</li>
</ul>

<h3 id="weekly-course-components">Weekly Course Components</h3>

<ul>
  <li><strong>In-person lecture and hands on demonstrations</strong>: Tuesdays, 10am-12pm starting January 28th</li>
  <li><strong>In-person practice sessions</strong>: Fridays at 10am-12pm starting on January 31st
    <ul>
      <li>This is a chance to review material from Tuesday and work on the weekly assignments (see below) with help from Dr. Chang, BSPC staff as well as fellow students. You are strongly encouraged to attend these sessions to get the most out of this course.</li>
    </ul>
  </li>
  <li><strong>Asynchronous Practice</strong>: 3-4 hours of asynchronous practice and reading. This will prepare you for the following week and allow you to apply knowledge to new examples. You will submit small weekly assignment to me so I can check your progress and adjust content if needed.</li>
</ul>

<h3 id="topics-and-links-to-weekly-materials">Topics and Links to weekly materials</h3>

<p>Note: Links to weekly lessons will be provided several days in advance in the table below. All course materials are living documents - some changes may be made after publishing to reflect our progress through the material.</p>

<table>
  <thead>
    <tr>
      <th>Link to Materials</th>
      <th>Topic</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><a href="/branched-docs-demo/site-v2/schedule/links-to-lessons.html#week-1">Week 1</a></td>
      <td>Introduction to the command line and logging into Biowulf</td>
    </tr>
    <tr>
      <td><a href="/branched-docs-demo/site-v2/schedule/links-to-lessons.html#week-2">Week 2</a></td>
      <td>Scientific software on Biowulf, quality control of sequence data, experimental design</td>
    </tr>
    <tr>
      <td><a href="/branched-docs-demo/site-v2/schedule/links-to-lessons.html#week-3">Week 3</a></td>
      <td>Reference genomes, theory and practice of mapping RNAseq reads to a reference, Mapping QC</td>
    </tr>
    <tr>
      <td><a href="/branched-docs-demo/site-v2/schedule/links-to-lessons.html#week-4">Week 4</a></td>
      <td>Theory and practice of counting RNAseq reads</td>
    </tr>
    <tr>
      <td><a href="/branched-docs-demo/site-v2/schedule/links-to-lessons.html#week-5">Week 5</a></td>
      <td>Automation of the RNAseq workflow, transition to RStudio for differential expression analyses</td>
    </tr>
    <tr>
      <td><a href="/branched-docs-demo/site-v2/schedule/links-to-lessons.html#week-6">Week 6</a></td>
      <td>Overview and prep for DEseq2 analysis pipeline, assessing sample quality, design formulas</td>
    </tr>
    <tr>
      <td><a href="/branched-docs-demo/site-v2/schedule/links-to-lessons.html#week-7">Week 7</a></td>
      <td>Hypothesis testing in DESeq2, summarizing and visualizing results</td>
    </tr>
    <tr>
      <td><a href="/branched-docs-demo/site-v2/schedule/links-to-lessons.html#week-8">Week 8</a></td>
      <td>Likelihood ratio test results, overenrichment analyses, functional class scoring</td>
    </tr>
    <tr>
      <td><a href="/branched-docs-demo/site-v2/schedule/links-to-lessons.html#week-9">Week 9</a></td>
      <td>Any material left from Week 8, course wrap-up, review of requested topics</td>
    </tr>
  </tbody>
</table>

<h3 id="course-communication">Course communication:</h3>
<ul>
  <li>For any content questions: We will be using the <code class="language-plaintext highlighter-rouge">rnaseq_jan2025</code> Slack channel under the Bioinformatics @NICHD workspace. Posting your questions in this channel is great, as multiple staff and fellow students will be able to respond. Also, you probably aren’t the only one with that question!</li>
  <li>For scheduling/personal concerns: Contact Dr. Chang by e-mail.</li>
</ul>

<h3 id="software-requirements">Software Requirements</h3>

<p><strong><em>Mac users:</em></strong></p>

<ul>
  <li>Plain text editor: TextEdit should be installed by default on Macs.</li>
</ul>

<p><strong><em>Windows users:</em></strong></p>

<ul>
  <li><a href="https://git-scm.com/download/win">GitBash</a></li>
  <li>Plain text editor: Microsoft Notepad should be installed by default on Windows.</li>
</ul>

<p><strong>Note about text plain text editors:</strong> A plain text editor is a program to edit text files such as a script that doesn’t interfere with formatting like a full word processor (like Word) would. The built-in text editors for each operating system are listed above.</p>

<p>As we progress in the course and your research, you may find that you want a “fancier” text editor that has more features for coding efficiently.</p>

<p>The following options are approved by NICHD IT but may need a license after a trial period:</p>
<ul>
  <li>For Macs: <a href="https://www.barebones.com/products/bbedit/index.html">BBEdit</a>, which has a lot more features for coding but needs a license to keep using all of those features after the trial period.</li>
  <li>For PCs: <a href="https://notepad-plus-plus.org/">NotePad++</a>, which likewise has many more features and is FREE!</li>
</ul>

<hr />

<h3 id="citation">Citation</h3>

<p>These materials were modified by E. Sally Chang at NICHD from the following citation:</p>

<blockquote>
  <p>Mary E. Piper, Meeta Mistry, Jihe Liu, William J. Gammerdinger, &amp; Radhika S. Khetani. (2022, January 10). hbctraining/Intro-to-rnaseq-hpc-salmon-flipped: Introduction to RNA-seq using Salmon Lessons from HCBC (first release). Zenodo. <a href="https://doi.org/10.5281/zenodo.5833880">https://doi.org/10.5281/zenodo.5833880</a>. RRID:SCR_025373.</p>
</blockquote>

<hr />

<p><em>The original materials developed by members of the teaching team at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>. These are open access materials distributed under the terms of the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.</em></p>

<ul>
  <li><em>Some materials used in these lessons were derived from work that is Copyright © Data Carpentry (<a href="http://datacarpentry.org/">http://datacarpentry.org/</a>). All Data Carpentry instructional material is made available under the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0).</em></li>
</ul>


      <footer class="site-footer">
        
          <span class="site-footer-owner"><a href="https://github.com/NICHD-BSPC/branched-docs-demo">branched-docs-demo</a> is maintained by <a href="https://github.com/NICHD-BSPC">NICHD-BSPC</a>.</span>
        
        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a>.</span>
      </footer>
    </main>
  </body>
</html>
