<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link rel="stylesheet" href="/branched-docs-demo/assets/css/style.css?v=37de6fdd01aa8b3215dd19a3556c476cb131dc09" media="screen" type="text/css">
    <link rel="stylesheet" href="/branched-docs-demo/assets/css/print.css" media="print" type="text/css">

    <!--[if lt IE 9]>
    <script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
    <![endif]-->

<!-- Begin Jekyll SEO tag v2.8.0 -->
<title>Data Management and Project Organization | Branched Docs Demo</title>
<meta name="generator" content="Jekyll v3.10.0" />
<meta property="og:title" content="Data Management and Project Organization" />
<meta name="author" content="Harvard HPC Staff, Modified by E. Sally Chang @ NICHD" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Example of building HTML docs hosted on github.io with subdirectories for selected branches" />
<meta property="og:description" content="Example of building HTML docs hosted on github.io with subdirectories for selected branches" />
<link rel="canonical" href="https://nichd-bspc.github.io/branched-docs-demo/lessons/wk1_lesson05_data_organization.html" />
<meta property="og:url" content="https://nichd-bspc.github.io/branched-docs-demo/lessons/wk1_lesson05_data_organization.html" />
<meta property="og:site_name" content="Branched Docs Demo" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2025-03-01T00:00:00+00:00" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="Data Management and Project Organization" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"BlogPosting","author":{"@type":"Person","name":"Harvard HPC Staff, Modified by E. Sally Chang @ NICHD"},"dateModified":"2025-03-01T00:00:00+00:00","datePublished":"2025-03-01T00:00:00+00:00","description":"Example of building HTML docs hosted on github.io with subdirectories for selected branches","headline":"Data Management and Project Organization","mainEntityOfPage":{"@type":"WebPage","@id":"https://nichd-bspc.github.io/branched-docs-demo/lessons/wk1_lesson05_data_organization.html"},"url":"https://nichd-bspc.github.io/branched-docs-demo/lessons/wk1_lesson05_data_organization.html"}</script>
<!-- End Jekyll SEO tag -->


    <!-- start custom head snippets, customize with your own _includes/head-custom.html file -->

<!-- Setup Google Analytics -->



<!-- You can set your favicon here -->
<!-- link rel="shortcut icon" type="image/x-icon" href="/branched-docs-demo/favicon.ico" -->

<!-- end custom head snippets -->

  </head>

  <body>
    <header>
      <div class="inner">
        <a href="https://nichd-bspc.github.io/branched-docs-demo/">
          <h1>Branched Docs Demo</h1>
        </a>
        <h2>Example of building HTML docs hosted on github.io with subdirectories for selected branches</h2>
        
          <a href="https://github.com/NICHD-BSPC/branched-docs-demo" class="button"><small>View project on</small> GitHub</a>
        
        
      </div>
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
          <h3 id="note">NOTE:</h3>
<p>To make names more generalized for the next course, <code class="language-plaintext highlighter-rouge">/data/Bspc-training/shared/rnaseq_jan2025</code> is now <code class="language-plaintext highlighter-rouge">/data/Bspc-training/shared/rnaseq_mov10</code> . Make sure to edit any scripts that refer to the shared data!</p>

<h2 id="learning-objectives">Learning Objectives</h2>

<ul>
  <li>Explain the need for data management.</li>
  <li>Describe strategies for managing your own projects.</li>
  <li>Describe the example RNA-seq experiment and its objectives.</li>
  <li>Build on CLI skills from this week’s lessons to set up a organized work directory</li>
</ul>

<h2 id="what-is-data-management">What is data management?</h2>

<p>In this course we teach you how to independently analyze your own sequencing data, so naturally we should begin with what what to do once you get the raw data back from the sequencing facility.
Obviously, we begin with the analysis!
Right?</p>

<p>Wrong.</p>

<p>One of the most important parts of research that involves large amounts of data is how best to manage it.
Once data is generated we tend to prioritize the analysis.
<strong>In the excitement to get a first look at new data, there are many important aspects that are often overlooked.</strong></p>

<p>Wait, don’t leave this page just yet!</p>

<p>We know that data management can be hard to get excited about.
However, <strong>ignoring it can be detrimental to your research.</strong> Here are just a few reasons <strong>why data management should matter to you</strong>:</p>

<ul>
  <li>It will make your life easier. It’s easier to analyze organized, and well documented data.</li>
  <li>Your future self will thank you. Managing well from the get-go means it’s easier to retrieve at a later date.</li>
  <li>Data is precious. Thinking ahead about things like storage means you reduce risk of losing it.</li>
  <li>Funding agencies are increasingly mandating that research projects are developed with a data management plan.</li>
</ul>

<h3 id="data-lifecycle">Data Lifecycle</h3>

<p>The data lifecycle displayed below, courtesy of the <a href="https://datamanagement.hms.harvard.edu/">HMS Data Management Working Group</a>, illustrates some things to consider beyond data creation and analysis.
Below, we discuss components of the lifecycle and how they apply to any NGS experiment.</p>

<p align="center">

<img src="../img/data-lifecycle-base.png" width="600" />

</p>

<p><em>Image acquired from the <a href="https://datamanagement.hms.harvard.edu/data-lifecycle">Harvard Biomedical Data Management Website</a></em></p>

<h3 id="plan-and-design">Plan and Design</h3>

<p>You should approach your sequencing project in a very similar way you do with any biological experiment, and ideally, begins with a good <strong>experimental design</strong>.
You want to think about experiment at the outset and collect appropriate samples such that you have enough statistical power to make the comparisons you need.
<strong>In a later lesson, we delve more into the details of planning and the experimental design considerations.</strong></p>

<p>Planning for your computational work is just as important as planning when working on the bench.
Every computational analysis you do is going to spawn many files and you will want to think about short-term storage options for your data and computational resources for analyzing it.</p>

<h3 id="collect-and-create">Collect and Create</h3>

<p>The next step is preparing samples as required.
During this stage it is important to keep track of how the experiment was performed, making sure to clearly document the source of starting materials and kits used.
It is also best practice to include any information about any small variations within the experiment (across the different samples being prepared) or any changes relative to standard experiment protocols.
This collection of information serves as the <strong>metadata of the experiment</strong> which will prove to be very useful during the analysis stage.</p>

<h3 id="analyze-and-collaborate---our-course-will-mostly-focus-on-this-portion-of-the-life-cycle">Analyze and Collaborate - our course will mostly focus on this portion of the life cycle!</h3>

<p>Once you have the sequencing data back from the sequencing facility, it’s time to analyze it.
The process of data analysis should be well documented to ensure reproducibility and also for ease of collaboration.
We will spend some more time on this component of the lifecycle later in class, as it applies to our dataset.</p>

<h3 id="evaluate-and-archive">Evaluate and Archive</h3>

<p>When the analysis is complete you will want to think about which files are most pertinent to keep.
Consider long-term storage options for your data that meet requirements of NIH, other funding agencies, and any guidelines from your institution.</p>

<h3 id="disseminate-and-share">Disseminate and share</h3>

<p>The results of your analysis will hopefully generate some exciting findings that will be beneficial to the scientific community.
At this stage in the lifecycle you rely on your previous steps of documentation to turn those notes into a clear and concise methods section of your manuscript.</p>

<h3 id="access-and-reuse">Access and Reuse</h3>

<p>In addition to sharing information on the analysis, you should plan for sharing the data.
It has become increasingly common for researchers to make their data available to others when they complete a study.
While a major reason for sharing is compliance (with journals or research funding organizations), there are also important research benefits including reproducibility and data sharing and reuse.</p>

<h2 id="the-dataset">The Dataset</h2>

<p>The dataset we are using for this workshop is part of a larger study described in <a href="http://www.ncbi.nlm.nih.gov/pubmed/25464849">Kenny PJ et al., <em>Cell Rep</em> 2014</a>.
The authors are investigating interactions between various genes involved in Fragile X syndrome, a disease of aberrant protein production, which results in cognitive impairment and autistic-like features.
<strong>The authors sought to show that RNA helicase MOV10 regulates the translation of RNAs involved in Fragile X syndrome.</strong></p>

<h3 id="raw-data">Raw data</h3>

<p>From this study we are using the <a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE50499">RNA-seq</a> data which is publicly available in the <a href="https://www.ncbi.nlm.nih.gov/sra/?term=SRP029367">Sequence Read Archive (SRA)</a>.</p>

<blockquote>
  <p><strong>NOTE:</strong> If you are interested in how to obtain publicly available sequence data from the SRA, the Harvard HPC staff have created a lesson about this (note that it is not customized for Biowulf) <a href="https://hbctraining.github.io/Accessing_public_genomic_data/lessons/downloading_from_SRA.html">linked here</a>.</p>
</blockquote>

<h3 id="metadata">Metadata</h3>

<p>In addition to the raw sequence data we also need to collect <strong>information about the data</strong>, also known as <strong>metadata</strong>.
We are usually quick to want to begin analysis of the sequence data (FASTQ files), but how useful is it if we know nothing about the samples that this sequence data originated from?</p>

<p>Some relevant metadata for our dataset is provided below:</p>

<ul>
  <li>The RNA was extracted from <strong>HEK293F cells</strong> that were transfected with a <strong>MOV10 transgene</strong>, <strong>MOV10 siRNA</strong>, or an <strong>irrelevant siRNA</strong>. (<em>For this course we won’t be using the MOV10 knock down samples.</em>)</li>
  <li>The libraries for this dataset are <strong>stranded</strong> and were generated using the standard Tru-seq prep kit (using the dUTP method).</li>
  <li>Sequencing was carried out on the <strong>Illumina HiSeq-2500</strong> and <strong>100bp single end</strong> reads were generated.</li>
  <li>The full dataset was sequenced to <strong>~40 million reads</strong> per sample, but for this workshop we will be looking at a small subset on chr1 (~300,000 reads/sample).</li>
  <li>For each group we have three replicates as described in the figure below.</li>
</ul>

<p align="center">

<img src="../img/exp_design.png" width="700" />

</p>

<h2 id="exercise-implementing-data-management-best-practices">Exercise: Implementing data management best practices</h2>

<p>Here, we implement some of those strategies to get ourselves setup before we begin with any analysis.</p>

<h3 id="are-you-logged-into-biowulf-on-a-compute-node-and-in-your-course-directory">Are you logged into Biowulf, on a compute node, and in your course directory?</h3>

<p>If not, first do that before proceeding!
As a reminder, the course directory is: <code class="language-plaintext highlighter-rouge">/data/Bspc-training/$USER</code></p>

<h3 id="planning-and-organization">Planning and organization</h3>

<p>For each experiment you work on and analyze data for, it is considered best practice to get organized by creating a planned storage space (directory structure).
We will start by creating a directory that we can use for the rest of the workshop.
First, make sure that you are in your home directory.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span><span class="nb">cd</span> /data/Bspc-training/<span class="nv">$USER</span>
</code></pre></div></div>

<p>From here, create the directory <code class="language-plaintext highlighter-rouge">rnaseq</code> and move into it.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span><span class="nb">mkdir </span>rnaseq
<span class="nv">$ </span><span class="nb">cd </span>rnaseq
</code></pre></div></div>

<p>Next, we will create a project directory and set up the following structure to keep our files organized.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>rnaseq
  ├── logs
  ├── meta
  ├── raw_data  
  ├── results
  └── scripts
</code></pre></div></div>

<p><em>This is a generic structure and can be tweaked based on personal preference and the analysis workflow.</em></p>

<ul>
  <li><code class="language-plaintext highlighter-rouge">logs</code>: to keep track of the commands run and the specific parameters used, but also to have a record of any standard output that is generated while running the command.</li>
  <li><code class="language-plaintext highlighter-rouge">meta</code>: for any information that describes the samples you are using, which we refer to as <a href="https://datamanagement.hms.harvard.edu/metadata-overview">metadata</a>.</li>
  <li><code class="language-plaintext highlighter-rouge">raw_data</code>: for any <strong>unmodified</strong> (raw) data obtained prior to computational analysis here, e.g. FASTQ files from the sequencing center. We strongly recommend leaving this directory unmodified through the analysis.</li>
  <li><code class="language-plaintext highlighter-rouge">results</code>: for output from the different tools you implement in your workflow. Create sub-folders specific to each tool/step of the workflow within this folder.</li>
  <li><code class="language-plaintext highlighter-rouge">scripts</code>: for scripts that you write and use to run analyses/workflow.</li>
</ul>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span><span class="nb">mkdir </span>logs meta raw_data results scripts
</code></pre></div></div>

<blockquote>
  <h4 id="file-naming-conventions">File naming conventions</h4>

  <p>Another aspect of staying organized is making sure that all the directories and filenames for an analysis are as consistent as possible.
You want to avoid names like <code class="language-plaintext highlighter-rouge">alignment1.bam</code>, and rather have names like <code class="language-plaintext highlighter-rouge">20170823_kd_rep1_gmap-1.4.bam</code> which provide a basic level of information about the file.
<a href="https://datamanagement.hms.harvard.edu/file-naming-conventions">This link</a> and <a href="http://www2.stat.duke.edu/~rcs46/lectures_2015/01-markdown-git/slides/naming-slides/naming-slides.pdf">this slideshow</a> provided by Harvard HPC have some good guidelines for file naming dos and don’ts.</p>
</blockquote>

<h3 id="documentation-worklog-and-readme-files">Documentation: Worklog and README files</h3>

<p>In your lab notebook, you likely keep track of the different reagents and kits used for a specific protocol.
Similarly, recording information about the tools used in the workflow is important for documenting your computational experiments.</p>

<p>Here are some things you might want to make a note of:</p>

<ul>
  <li>
    <p><strong>Make note of the software you use.</strong> Do your research and find out what tools are best for the data you are working with.
Don’t just work with tools that you are able to easily install.</p>
  </li>
  <li>
    <p><strong>Keep track of software versions.</strong> Keep up with the literature and make sure you are using the most up-to-date versions.</p>
  </li>
  <li>
    <p><strong>Record information on parameters used and summary statistics</strong> at every step (e.g., how many adapters were removed, how many reads did not align)</p>
  </li>
  <li>
    <p><strong>Test datasets:</strong> A general rule of thumb is to test on a single sample or a subset of the data before running your entire dataset through.
This will allow you to debug quicker and give you a chance to also get a feel for the tool and the different parameters.</p>
  </li>
  <li>
    <p><strong>Error and log messages</strong>: Different tools have different ways of reporting log messages to the terminal.
You might have to experiment a bit to figure out what output to capture.
You can redirect standard output with the <code class="language-plaintext highlighter-rouge">&gt;</code> symbol which is equivalent to <code class="language-plaintext highlighter-rouge">1&gt; (standard out)</code>; other tools might require you to use <code class="language-plaintext highlighter-rouge">2&gt;</code> to re-direct the <code class="language-plaintext highlighter-rouge">standard error</code> instead.</p>
  </li>
</ul>

<h4 id="creating-a-worklog">Creating a Worklog</h4>

<p>The BSPC uses a <em>Work Log</em> for each project that they work on.
This is a file that lives in the top directory for a project (i.e. <code class="language-plaintext highlighter-rouge">rnaseq</code> in our case), and has the following attributes:</p>

<ul>
  <li>Named <code class="language-plaintext highlighter-rouge">WORKLOG.rst</code> to make it easy to find across projects</li>
  <li>Use ReStructured Text format (<code class="language-plaintext highlighter-rouge">.rst</code> files), because this allows for more flexible documentation and parsing down the line</li>
  <li>Start each entry with a date, in YYYY-MM-DD format, underlined with hyphens</li>
  <li>Then a blank line</li>
  <li>Then your username prefixed with an \@</li>
  <li>Then arbitrary text describing what you did that day.</li>
  <li>Keeps everything in chronological order, so the latest entry is at the bottom</li>
</ul>

<p>For example, the top of my Work Log for the <code class="language-plaintext highlighter-rouge">rnaseq</code> project looks like:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>2024-10-03
<span class="nt">----------</span>

@changes

- Created directory structure, README, and Work Log
</code></pre></div></div>

<p>The audience is you (or another lab member) 2 years from now when everyone has forgotten about the project and now needs to start working on it again.
As you are writing, it will feel like you are writing too much or adding way too much detail.
You are not.
<strong>The future you will thank you.</strong></p>

<p>Feel free to keep an informal, conversational tone.
This is not a public document.
<em>But do get in the habit of adding a note each time you work on a project!</em></p>

<h4 id="readme-files">README files</h4>

<p>After setting up the directory structure it is useful to have a <a href="https://datamanagement.hms.harvard.edu/readme-files"><strong>README file</strong></a> <strong>within your project directory</strong>.
This is a plain text file containing a short summary about the project and a description of the files/directories found within it.
An example README is shown below.
It can also be helpful to include a README within each sub-directory with any information pertaining to the analysis.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## README ##
## This directory contains data generated during the Introduction to RNA-seq workshop
## Date: 

There are five subdirectories in this directory:

raw_data : contains raw data
meta:  contains...
logs:
results:
scripts:
</code></pre></div></div>

<h3 id="moving-data-into-our-new-raw_data-folder">Moving data into our new raw_data folder</h3>

<p>Let’s populate the <code class="language-plaintext highlighter-rouge">rnaseq/</code> project with some data.
We actually have the required files that we were manipulating in our previous few lessons.
Let’s <code class="language-plaintext highlighter-rouge">mv</code> these files into our raw_data directory.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span><span class="nb">mv </span>unix_lesson/raw_fastq/<span class="k">*</span>subset.fq raw_data/
</code></pre></div></div>

<p>If you were to run the <code class="language-plaintext highlighter-rouge">tree</code> command from inside the <code class="language-plaintext highlighter-rouge">rnaseq</code> directory it should now look like this:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>rnaseq
  ├── logs
  ├── meta
  ├── raw_data
  │   ├── Irrel_kd_1.subset.fq
  │   ├── Irrel_kd_2.subset.fq
  │   ├── Irrel_kd_3.subset.fq
  │   ├── Mov10_oe_1.subset.fq
  │   ├── Mov10_oe_2.subset.fq
  │   └── Mov10_oe_3.subset.fq
  ├── README.txt
  ├── results
  └── scripts
</code></pre></div></div>

<h3 id="assignment"><strong>Assignment:</strong></h3>

<ol>
  <li>Take a moment to create a README for the <code class="language-plaintext highlighter-rouge">rnaseq/</code> folder (hint: use <code class="language-plaintext highlighter-rouge">vim</code> to create the file). Give a short description of the project and brief descriptions of the types of files you will be storing within each of the sub-directories.</li>
  <li>Use the same strategy to create a <code class="language-plaintext highlighter-rouge">WORKLOG.rst file</code> and add your first entry about what you just accomplished.</li>
</ol>

<p><strong>Note</strong>: Message me on Slack when you’ve finished this lesson. I’ll look into your <code class="language-plaintext highlighter-rouge">/rnaseq</code> directory to see that you created the assigned directory structure and these two text files!</p>

<hr />

<p><em>This lesson has been developed by members of the teaching team at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>. These are open access materials distributed under the terms of the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.</em></p>

<ul>
  <li><em>The materials used in this lesson were derived from work that is Copyright © Data Carpentry (<a href="http://datacarpentry.org/">http://datacarpentry.org/</a>). All Data Carpentry instructional material is made available under the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0).</em></li>
  <li><em>Adapted from the lesson by Tracy Teal. Original contributors: Paul Wilson, Milad Fatenejad, Sasha Wood and Radhika Khetani for Software Carpentry (<a href="http://software-carpentry.org/">http://software-carpentry.org/</a>)</em></li>
</ul>

        </section>

        <aside id="sidebar">
          

          
            <p class="repo-owner"><a href="https://github.com/NICHD-BSPC/branched-docs-demo">branched-docs-demo</a> is maintained by <a href="https://github.com/NICHD-BSPC">NICHD-BSPC</a>.</p>
          

          <p>This page was generated by <a href="https://pages.github.com">GitHub Pages</a>.</p>
        </aside>
      </div>
    </div>

  </body>
</html>
