<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="UTF-8">

<!-- Begin Jekyll SEO tag v2.8.0 -->
<title>MultiQC | Branched Docs Demo</title>
<meta name="generator" content="Jekyll v3.10.0" />
<meta property="og:title" content="MultiQC" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Example of building HTML docs hosted on github.io with subdirectories for selected branches" />
<meta property="og:description" content="Example of building HTML docs hosted on github.io with subdirectories for selected branches" />
<link rel="canonical" href="https://nichd-bspc.github.io/branched-docs-demo/site-v2/lessons/wk3_lesson03_multiqc.html" />
<meta property="og:url" content="https://nichd-bspc.github.io/branched-docs-demo/site-v2/lessons/wk3_lesson03_multiqc.html" />
<meta property="og:site_name" content="Branched Docs Demo" />
<meta property="og:type" content="website" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="MultiQC" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"WebPage","description":"Example of building HTML docs hosted on github.io with subdirectories for selected branches","headline":"MultiQC","url":"https://nichd-bspc.github.io/branched-docs-demo/site-v2/lessons/wk3_lesson03_multiqc.html"}</script>
<!-- End Jekyll SEO tag -->

    <link rel="preconnect" href="https://fonts.gstatic.com">
    <link rel="preload" href="https://fonts.googleapis.com/css?family=Open+Sans:400,700&display=swap" as="style" type="text/css" crossorigin>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="theme-color" content="#157878">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    <link rel="stylesheet" href="/branched-docs-demo/site-v2/assets/css/style.css?v=1a7a7ab75167ab0a3d6257c2d7411dd1014d090f">
    <!-- start custom head snippets, customize with your own _includes/head-custom.html file -->

<!-- Setup Google Analytics -->



<!-- You can set your favicon here -->
<!-- link rel="shortcut icon" type="image/x-icon" href="/branched-docs-demo/site-v2/favicon.ico" -->

<!-- end custom head snippets -->

  </head>
  <body>
    <a id="skip-to-content" href="#content">Skip to the content.</a>

    <header class="page-header" role="banner">
      <h1 class="project-name">MultiQC</h1>
      <h2 class="project-tagline">Example of building HTML docs hosted on github.io with subdirectories for selected branches</h2>
      
        <a href="https://github.com/NICHD-BSPC/branched-docs-demo" class="btn">View on GitHub</a>
      
      
    </header>

    <main id="content" class="main-content" role="main">
      <p>Approximate time: 30 minutes</p>

<h3 id="note">NOTE:</h3>
<p>To make names more generalized for the next course, <code class="language-plaintext highlighter-rouge">/data/Bspc-training/shared/rnaseq_jan2025</code> is now <code class="language-plaintext highlighter-rouge">/data/Bspc-training/shared/rnaseq_mov10</code> . Make sure to edit any scripts that refer to the shared data!</p>

<h2 id="learning-objectives">Learning Objectives</h2>

<ul>
  <li>Run the multiQC tool to gather QC metrics from multiple tools for all samples</li>
  <li>Assess and compare QC metrics among samples</li>
</ul>

<h2 id="documenting-results-and-gathering-qc-metrics">Documenting results and gathering QC metrics</h2>

<p>As you go through the RNA-seq workflow (or any data analysis workflow), it is important to document the parameters you used for running the analysis. In addition, it is also very important to document the metrics/results at every step. Careful evaluation of metrics is a form of QC, and it will enable you to identify any issues with the data and/or the parameters you are using, as well as alert you to the presence of contamination or systematic biases, etc.</p>

<p>There are several metrics you can evaluate in the RNA-seq workflow. Below are 3 important ones that you should keep track of for each sample:</p>

<ul>
  <li>number of raw reads</li>
  <li>percentage of reads aligned to genome</li>
  <li>percentage of reads associated with genes</li>
</ul>

<p>An important QC step is to make sure that these metrics are consistent across the samples for a given experiment, and any outliers should be investigated further.</p>

<p>Manually tracking these metrics is tedious and error-prone. Many tools can help you with the documentation and QC assessment, some of which also have really nice visualizations to easily identify any issues, e.g. <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>, <a href="http://qualimap.bioinfo.cipf.es/doc_html/index.html">Qualimap</a>, <a href="http://multiqc.info/">MultiQC</a>. Some of these tools tend to focus on a single sample at a time, and on QC for a specific step in the workflow. MultiQC, on the other hand, is able to make a report from the output of many different tools (for RNA-seq analysis and other NGS workflows) and it is able to combine the information for multiple samples.</p>

<h3 id="tracking-and-aggregating-results-from-workflow-tools-with-multiqc">Tracking and aggregating results from workflow tools with <em>MultiQC</em></h3>

<p>In this lesson, we will be using MultiQC to aggregate results from several tools and generates a single HTML report with plots to visualize and compare QC metrics between the samples.</p>

<p>MultiQC can generate this report from 96 different bioinformatics tools, and these tools span various NGS analyses, e.g., basic QC, RNA-seq, ChIP-seq, variant calling, genome annotation, etc. We are going to use it to aggregate information from the results of <a href="http://multiqc.info/docs/#fastqc">FastQC</a>, <a href="http://multiqc.info/docs/#star">STAR</a>, and <a href="http://multiqc.info/docs/#qualimap">Qualimap</a>. MultiQC can parse the information from <strong>specific output files</strong> of these tools.</p>

<p>Start by creating a directory for our output called <code class="language-plaintext highlighter-rouge">multiqc_report</code>:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span><span class="nb">cd</span> /data/Bspc-training/<span class="nv">$USER</span>/rnaseq <span class="c">#if needed</span>

<span class="nv">$ </span><span class="nb">mkdir </span>results/multiqc_report
</code></pre></div></div>

<p>Then navigate into that directory:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span><span class="nb">cd </span>results/multiqc_report
</code></pre></div></div>

<p>Next, load MultiQC. The default version is fine but make sure to note what version it is!</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>module load multiqc <span class="c">#loads v1.22 right now </span>
</code></pre></div></div>

<hr />

<p>We are going to run MultiQC on the following 4 outputs from our workflow:</p>

<ul>
  <li><code class="language-plaintext highlighter-rouge">.zip</code> files from FastQC</li>
  <li><code class="language-plaintext highlighter-rouge">.Log.final.out</code> files from STAR</li>
  <li>files from Qualimap</li>
</ul>

<p>To create a more meaningful report to look at we thought it best to run MultiQC on the full dataset instead of the subset we have been working with so far. We have run each of the tools mentioned above on the full dataset and stored the result in the directory <code class="language-plaintext highlighter-rouge">/data/Bspc-training/shared/rnaseq_jan2025/results_for_multiqc</code>. We will point to these files as input for our MultiQC analysis.</p>

<p>To run MultiQC, we can provide it two inputs at a minimum:</p>

<ol>
  <li>a name for our output report and folder</li>
  <li>the paths to our results files</li>
</ol>

<blockquote>
  <p><strong>NOTE:</strong> MultiQC has additional parameters we could include; use <code class="language-plaintext highlighter-rouge">multiqc -h</code> to find out more.</p>
</blockquote>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>multiqc <span class="nt">-n</span> multiqc_report_rnaseq /data/Bspc-training/shared/rnaseq_jan2025/results_for_multiqc/fastqc/<span class="k">*</span>zip /data/Bspc-training/shared/rnaseq_jan2025/results_for_multiqc/STAR/<span class="k">*</span>Log.final.out /data/Bspc-training/shared/rnaseq_jan2025/results_for_multiqc/qualimap/<span class="k">*</span>
</code></pre></div></div>

<blockquote>
  <p><strong>NOTE</strong>: You will see the progress of analysis printed out on the terminal as the tool runs. If you want to save this output into a log file (for future reference), you can use <code class="language-plaintext highlighter-rouge">2&gt;</code> operator to redirect it to a file. For example, at the end of script, add <code class="language-plaintext highlighter-rouge">2&gt; log.txt</code>. <code class="language-plaintext highlighter-rouge">2&gt;</code>redirects the output of so-called standard error.</p>
</blockquote>

<p>It takes a couple of minutes to generate the MultiQC report. The report provides nice visualizations across samples, which is very useful to determine consistency and to identify problematic samples.</p>

<p>The output of MultiQC is one HTML file (<code class="language-plaintext highlighter-rouge">multiqc_report_rnaseq.html</code>) and a data folder. Transfer the interactive HTML report over to your laptop using <code class="language-plaintext highlighter-rouge">scp</code> or the locally mounted server option. You can also download the output HTML file <a href="/branched-docs-demo/site-v2/data/multiqc_report_rnaseq_ALL_samples.html">here</a> to follow along.</p>

<h2 id="assessing-the-quality-control-metrics">Assessing the quality control metrics</h2>

<p>The main metrics to explore first are:</p>

<ul>
  <li>number of raw reads or total reads</li>
  <li>percentage of reads aligned to genome</li>
  <li>percentage of reads associated with genes</li>
</ul>

<blockquote>
  <p>Note: If you don’t see exact columns as ours, you may need to configure the columns, which is a button just underneath the ‘General Statistics’ heading.</p>
</blockquote>

<p align="center">

<img src="../img/multiqc_table.png" width="750" />

</p>

<p>Using <code class="language-plaintext highlighter-rouge">Configure Columns</code> button, we are going to choose the following columns:</p>

<p align="center">

<img src="../img/multiqc_columns.png" width="600" />

</p>

<p>In the above image, the description column is helpful in interpreting the table. Upon perusal of the table, we can see input from FastQC, STAR, and Qualimap. For example, the total number of raw reads is given in the <code class="language-plaintext highlighter-rouge">M Seqs</code> column on the far right of the table.</p>

<p>STAR provides information about <em>uniquely mapping reads</em> in the <code class="language-plaintext highlighter-rouge">%Aligned</code> column. A good quality sample will have <strong>at least 75% of the reads uniquely mapped</strong>. Once the value starts to drop below 60%, it’s advisable to start troubleshooting. Low number of uniquely mapping reads means that more reads are mapped to multiple locations.</p>

<p>The ‘STAR: Alignment Scores’ plot visually represents this mapping information. The % uniquely mapped, multimapped, and unmapped reads can be easily compared between samples to get a nice overview of the quality of the samples.</p>

<p align="center">

<img src="../img/multiqc_alignment_scores.png" width="600" />

</p>

<blockquote>
  <p>NOTE: The thresholds suggested above will vary depending on the organism that you are working with. Much of what is discussed here is in the context of working with human or mouse data. For example, 75% of mapped reads holds true only if the genome is good or mature. For badly assembled genomes, we may not observe a high mapping rate, even if the actual sequences from the sample are good.</p>
</blockquote>

<h3 id="complexity">Complexity</h3>

<p>The complexity of the RNA-seq library can be explored with the <code class="language-plaintext highlighter-rouge">%Dups</code> column. If a large percentage of the library is duplicated, then this could indicate a library of either low complexity or over-amplification. If huge differences of <code class="language-plaintext highlighter-rouge">%Dups</code> exist between samples, this may lead to biases in the data, such as different %GC content.</p>

<h3 id="exploring-biases">Exploring biases</h3>

<p>Within this report, we can also explore the bias metrics output by Qualimap and FastQC. The <code class="language-plaintext highlighter-rouge">5'-3' bias</code> column denotes whether our data has any 5’ or 3’ biases. These biases could be due to RNA degradation or different sample preparation techniques. Generally, we should explore our data more if we have biases approaching 0.5 or 2.</p>

<p>The transcript position plot can also help identify 5’ or 3’ bias, in addition to other coverage issues. We generally expect roughly even coverage.</p>

<p align="center">

<img src="../img/multiqc_coverage_profile.png" width="600" />

</p>

<p>In addition, we can see whether our different samples have differences in <code class="language-plaintext highlighter-rouge">%GC</code> column. GC bias could be caused by low-complexity libraries, differences in amplification, or library-specific issues. We expect to observe similar GC content aross samples.</p>

<h3 id="contamination">Contamination</h3>

<p>We can also identify possible contamination of our samples by inspecting the percentage of reads that are exonic, intronic or intergenic. High levels of intergenic reads is indicative of DNA contamination (&gt;30%). Also, if polyA selection of messenger RNAs was performed in library preparation, then high percentages of intronic reads would also be concerning.</p>

<p align="center">

<img src="../img/qualimap_genomic_origin.png" width="600" />

</p>

<p>Generally speaking, in a good library, we expect over 60% of reads to be mapped to exons for mouse or human organisms. For other organisms, the percentage depends on how well the genome is annotated.</p>

<h2 id="assignment">Assignment</h2>

<p>Examine the MultiQC report and in a <code class="language-plaintext highlighter-rouge">multiqc_notes.txt</code> file, briefly note two observations you can make about our data now that we can view all six files side-by-side.</p>


      <footer class="site-footer">
        
          <span class="site-footer-owner"><a href="https://github.com/NICHD-BSPC/branched-docs-demo">branched-docs-demo</a> is maintained by <a href="https://github.com/NICHD-BSPC">NICHD-BSPC</a>.</span>
        
        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a>.</span>
      </footer>
    </main>
  </body>
</html>
